### PR TITLE
7.2/59090 symfony security

### PR DESCRIPTION
**Pull request by rnuernberg**

### CODE REVIEW

File: src/CoreBundle/private/library/classes/dbobjects/TCMSContentBoxItem.class.php [hunk 1]
```diff
@@ -78,16 +80,16 @@ public function loadMenuItems()
 
             $oCMSModuleList = TdbCmsModuleList::GetList($query, $this->iLanguageId);
             while ($oCMSModule = $oCMSModuleList->Next()) {
-                $tableInUserGroup = $activeUser->oAccessManager->user->IsInGroups($oCMSModule->fieldCmsUsergroupId);
-                if ($tableInUserGroup) {
-                    $oMenuItem = new TCMSMenuItem_Module();
-                    $oMenuItem->SetData($oCMSModule->sqlData);
+                if (false === $securityHelper->isGranted(CmsPermissionAttributeConstants::ACCESS, $oCMSModule)) {
```
Comment (bestform) [inline src/CoreBundle/private/library/classes/dbobjects/TCMSContentBoxItem.class.php:83]: Can this method handle a CMSModule as input?

File: src/CoreBundle/private/library/classes/dbobjects/TCMSUser.class.php [hunk 2]
```diff
@@ -412,37 +268,18 @@ public function GetCurrentEditLanguageObject()
...
-    public function _LoadAccessManager()
-    {
-        if (!is_null($this->id)) {
-            $this->oAccessManager = new TAccessManager();
-            $this->oAccessManager->InitFromObject($this);
-        }
+        $sessionService->setCurrentEditLanguageIso6391($language);
     }
 
     /**
```
Comment (bestform) [inline src/CoreBundle/private/library/classes/dbobjects/TCMSUser.class.php:285]: That's a lot of public stuff vanishing. I understand why, but at this level we need to drop the idea that this could be version 7.2. This clearly needs to be version 8.0

File: src/CoreBundle/private/modules/CMSInterface/CMSInterface.class.php [hunk 3]
```diff
@@ -25,10 +28,16 @@ public function Execute()
...
-        $aGroupListUser = $oGlobal->oUser->oAccessManager->user->groups->GetGroupListAsArray();
-        $aGroupListUser = TTools::MysqlRealEscapeArray($aGroupListUser);
+        /** @var SecurityHelperAccess $securityHelper */
+        $securityHelper = ServiceLocator::get(SecurityHelperAccess::class);
+
+        $groups = $securityHelper->getUser()?->getGroups();
+        $aGroupListUser = TTools::MysqlRealEscapeArray($groups);
         $sGroupSQL = '';
         foreach ($aGroupListUser as $sGroupId => $sGroupSystemName) {
+            if ('-' === $sGroupId) {
```
Comment (bestform) [inline src/CoreBundle/private/modules/CMSInterface/CMSInterface.class.php:38]: Why was this added?

File: src/CoreBundle/private/modules/CMSModulePageTreeRPC/CMSModulePageTreeRPC.class.php [hunk 4]
```diff
@@ -99,23 +101,24 @@ public function getChildren()
...
+        if ('' !== $sPortalList) {
+            $query .= ' WHERE `id` NOT IN ('.$sPortalList.')';
         }
+        $portalMainNodes = $this->getDatabaseConnection()->fetchAllAssociative($query);
+        $aPortalExcludeList = array_map(fn($row) => $this->getDatabaseConnection()->quote($row['main_node_tree']), $portalMainNodes);
 
         $sPortalCondition = '';
         if (count($aPortalExcludeList) > 0) {
-            $sPortalCondition .= " AND T.`id` NOT IN ('".implode("','", $aPortalExcludeList)."')";
+            $sPortalCondition .= " AND T.`id` NOT IN ('".implode(', ', $aPortalExcludeList)."')";
```
Comment (bestform) [inline src/CoreBundle/private/modules/CMSModulePageTreeRPC/CMSModulePageTreeRPC.class.php:121]: Is it ok to drop the quotes from the query here?

File: src/CoreBundle/private/modules/MTHeader/views/breadcrumb.view.php [hunk 5]
```diff
@@ -1,5 +1,4 @@
 <?php
-if (isset($data['oUser'])) {
```
Comment (bestform) [inline src/CoreBundle/private/modules/MTHeader/views/breadcrumb.view.php:2]: Why is this deleted?

File: src/CoreBundle/private/modules/MTTableEditor/MTTableEditor.class.php [hunk 6]
```diff
@@ -196,9 +199,10 @@ public function Init()
             $bUserHasEditRight = $this->oTableManager->oTableEditor->AllowEdit();
 
             if (!$bIsReadOnlyRequest && ((!$bUserHasEditRight && !$bIsInsert && !$this->bIsReadOnlyMode) || ($this->bIsReadOnlyMode && !$bUserHasReadOnlyRight))) {
-                $oCMSUser = TCMSUser::GetActiveUser();
-                $oCMSUser->Logout();
-                $this->controller->HeaderURLRedirect(PATH_CMS_CONTROLLER);
+                /** @var RouterInterface $router */
+                $router = ServiceLocator::get('router');
+                $logout = $router->generate('app_logout');
+                $this->controller->HeaderURLRedirect($logout);                $this->controller->HeaderURLRedirect($logout);
```
Comment (bestform) [inline src/CoreBundle/private/modules/MTTableEditor/MTTableEditor.class.php:205]: We do not need to do this twice. ;)

File: src/SecurityBundle/src/CmsUser/CmsUserModel.php [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+        readonly private string $email,
+        readonly private string $cmsLanguageId,
+        readonly private array $availableLanguagesIsoCodes,
+        readonly private ?string $currentEditLanguageIsoCode,
+        readonly private array $availableEditLanguages,
+        readonly private ?string $password = null,
+        readonly private array $roles = [],
+        readonly private array $rights = [],
+        readonly private array $groups = [],
+        readonly private array $portals = []
```
Comment (bestform) [inline src/SecurityBundle/src/CmsUser/CmsUserModel.php:26]: Why make those private with getters, if they could be public? They are readonly.

**Review by bestform — CHANGES_REQUESTED** [review]

File: docs/backend-permissions.md [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+If you want to allow access to an object for only some roles, groups or rights, that object should implement `RestrictedByCmsRoleInterface`, 
+`RestrictedByCmsGroupInterface` or `RestrictedByCmsRightInterface` respectively.
+
+Objects extending `TCMSRecord` will be voted on by the `CmsTableObjectVoter`. In addition, you can check access to a table by passing
+the table name to `isGranted()`.
+
+Relevant permissions can be found in `CmsPermissionAttributeConstants`.
+
+If you need access to the symfony security component in a place where injection is not possible, you can access it using the 
+`SecurityHelperAccess` via the DependencyFinder service.
```
[the remark this answers is not in the document]
Comment (rnuernberg) [reply docs/backend-permissions.md:34]: i did

File: src/CmsBackendBundle/src/ChameleonSystemCmsBackendBundle.php [hunk 9]
```diff
@@ -0,0 +1,13 @@
+<?php
+
+namespace ChameleonSystem\CmsBackendBundle;
+
+use Symfony\Component\HttpKernel\Bundle\Bundle;
+
+class ChameleonSystemCmsBackendBundle extends Bundle
+{
+    public function getPath(): string
```
[the remark this answers is not in the document]
Comment (rnuernberg) [reply src/CmsBackendBundle/src/ChameleonSystemCmsBackendBundle.php:9]: I am 
image

File: src/CoreBundle/Bridge/Chameleon/Module/Sidebar/CustomMenuItemProvider.php [hunk 10]
```diff
@@ -34,16 +39,14 @@ public function createMenuItem(\TdbCmsMenuItem $menuItem): ?MenuItem
 
     private function isItemAccessAllowed(\TdbCmsMenuCustomItem $customItem): bool
     {
-        $activeUser = \TCMSUser::GetActiveUser();
-        if (null === $activeUser) {
+        /** @var SecurityHelperAccess $securityHelper */
+        $securityHelper = ServiceLocator::get(SecurityHelperAccess::class);
+        if (false === $securityHelper->isGranted('ROLE_CMS_USER')) {
```
[the remark this answers is not in the document]
Comment (rnuernberg) [reply src/CoreBundle/Bridge/Chameleon/Module/Sidebar/CustomMenuItemProvider.php:44]: It does make it easier to search for uses later... added const for ROLE_CMS_ADMIN and ROLE_CMS_USER

File: src/CoreBundle/CoreEvents.php [hunk 11]
```diff
@@ -18,8 +18,17 @@ final class CoreEvents
 
     const GLOBAL_RESOURCE_COLLECTION_COLLECTED_JAVASCRIPT = 'chameleon_system_core.resource_collection_collected.javascript';
 
+    /**
+     * @deprecated since 7.2 - use symfony LoginSuccessEvent
+     */
```
[the remark this answers is not in the document]
Comment (rnuernberg) [reply src/CoreBundle/CoreEvents.php:23]: removed the events

File: src/CoreBundle/DataAccess/DataAccessCmsMasterPagedefCacheDecorator.php [hunk 12]
```diff
@@ -67,12 +69,16 @@ public function get(string $id): ?CmsMasterPagdef
...
+        $securityHelper = ServiceLocator::get(SecurityHelperAccess::class);
+        $security = $securityHelper->getSecurity();
+
         $cacheKey = array(
             'type' => 'pagedefdata',
             'pagedef' => $pagedef,
             'requestMasterPageDef' => $this->inputFilterUtil->getFilteredInput('__masterPageDef', false),
             'isTemplateEngineMode' => $this->requestInfoService->isCmsTemplateEngineEditMode(),
-            'cmsuserdefined' => TGlobal::CMSUserDefined(),
+            'cmsuserdefined' => null !== $security->getToken() && $securityHelper->isGranted('ROLE_CMS_USER'),
```
[the remark this answers is not in the document]
Comment (rnuernberg) [reply src/CoreBundle/DataAccess/DataAccessCmsMasterPagedefCacheDecorator.php:81]: the token is null if the user is not logged in. The original security helper from symfony does not check for this and explodes hard. Our helper does - but I rather be safe and explicit here (as this code is called from frontend and backend)

File: src/CoreBundle/DataAccess/DataAccessCmsTblConf.php [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+        $tableExists = $this->connection->fetchOne(
+            'SELECT EXISTS(SELECT 1 FROM `cms_tbl_conf` WHERE `name` = :tableName)',
+            ['tableName' => $tableName]
+        );
+
+        if (false === $tableExists) {
+            return false;
+        }
+
+        return 1 === (int)$tableExists;
```
[the remark this answers is not in the document]
Comment (rnuernberg) [reply src/CoreBundle/DataAccess/DataAccessCmsTblConf.php:39]: fair

File: src/CoreBundle/DataAccess/DataAccessCmsTblConf.php [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+                INNER JOIN `cms_tbl_conf` ON %1$s.`source_id` = `cms_tbl_conf`.`id`
+                     WHERE `cms_tbl_conf`.`name` = :tableName',
+            $this->connection->quoteIdentifier($permissionTable)
+        );
+        return array_map(
+            static fn(array $row) => $row['target_id'],
+            $this->connection->fetchAllAssociative(
+                $query,
+                ['tableName' => $tableName]
+            )
```
[the remark this answers is not in the document]
Comment (rnuernberg) [reply src/CoreBundle/DataAccess/DataAccessCmsTblConf.php:63]: much better

File: src/CoreBundle/DataAccess/DataAccessCmsTblConfInterface.php [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+{
+    public const PERMISSION_MAPPING = [
+        CmsPermissionAttributeConstants::TABLE_EDITOR_NEW => 'cms_role_mlt',
+        CmsPermissionAttributeConstants::TABLE_EDITOR_EDIT => 'cms_role1_mlt',
+        CmsPermissionAttributeConstants::TABLE_EDITOR_DELETE => 'cms_role2_mlt',
+        CmsPermissionAttributeConstants::TABLE_EDITOR_EDIT_ALL => 'cms_role3_mlt',
+        CmsPermissionAttributeConstants::TABLE_EDITOR_ACCESS_ALL => 'cms_role6_mlt',
+        CmsPermissionAttributeConstants::TABLE_EDITOR_NEW_LANGUAGE => 'cms_role4_mlt',
+        //CmsPermissionAttributeConstants::WORKFLOW_PUBLISH => 'cms_role5_mlt',
+        //CmsPermissionAttributeConstants::VERSIONING => 'cms_role7_mlt',
```
[the remark this answers is not in the document]
Comment (rnuernberg) [reply src/CoreBundle/DataAccess/DataAccessCmsTblConfInterface.php:18]: The tables still exists - but the permission is not used anywhere (and per default disabled). But since they are still there, I now added them as constants

File: src/CoreBundle/EventListener/BackendBreadcrumbListener.php [hunk 16]
```diff
@@ -15,7 +15,9 @@
 use ChameleonSystem\CoreBundle\Service\RequestInfoServiceInterface;
 use ChameleonSystem\CoreBundle\Util\InputFilterUtilInterface;
 use Symfony\Component\HttpFoundation\RequestStack;
+use Symfony\Component\HttpKernel\Event\ControllerEvent;
 use Symfony\Component\HttpKernel\Event\RequestEvent;
+use Symfony\Component\HttpKernel\Event\ViewEvent;
```
[the remark this answers is not in the document]
Comment (rnuernberg) [reply src/CoreBundle/EventListener/BackendBreadcrumbListener.php:20]: removed

File: src/CoreBundle/Routing/ChameleonFrontendRouter.php [hunk 17]
```diff
@@ -104,7 +104,7 @@ private function isPagedefParamRoute(Request $request, $pagedef)
         if (null === $pagedef) {
             return false;
         }
-        if (PATH_CUSTOMER_FRAMEWORK_CONTROLLER !== $request->getPathInfo()) {
```
[the remark this answers is not in the document]
Comment (rnuernberg) [reply src/CoreBundle/Routing/ChameleonFrontendRouter.php:107]: it was changed to `PATH_CMS_CONTROLLER_FRONTEND` - I need a path within the cms firewall

File: src/CoreBundle/Service/Initializer/PortalDomainServiceInitializer.php [hunk 18]
```diff
@@ -108,7 +110,12 @@ private function determinePortalAndDomainDefault(Request $request, PortalDomainS
 
         $sName = $request->getHost();
         $sRelativePath = $request->getPathInfo();
-        $isUserSignedInToBackend = \TCMSUser::CMSUserDefined();
+        $isUserSignedInToBackend = false;
+        if ($request->hasSession()) {
```
[the remark this answers is not in the document]
Comment (rnuernberg) [reply src/CoreBundle/Service/Initializer/PortalDomainServiceInitializer.php:114]: the symfony helper will blow up if you call it without a session. Our wrapper does not  - but I wanted to be sure

File: src/CoreBundle/Service/LanguageService.php [hunk 19]
```diff
@@ -252,6 +253,7 @@ public function getActiveLocale()
      */
     public function getActiveEditLanguage()
     {
+        // todo: the language service is used in the backensession - so we cannot use the backend session here.
```
[the remark this answers is not in the document]
Comment (rnuernberg) [reply src/CoreBundle/Service/LanguageService.php:256]: jup, I removed the method

File: src/CoreBundle/UniversalUploader/Controller/UploaderController.php [hunk 20]
```diff
@@ -72,13 +77,7 @@ public function __invoke()
...
     private function isBackendUserAuthenticated(): bool
     {
-        $user = \TdbCmsUser::GetActiveUser();
-
-        if (null === $user || false === $user->bLoggedIn) {
-            return false;
-        }
-
-        return true;
+        return $this->security->isGranted('ROLE_CMS_USER');
```
[the remark this answers is not in the document]
Comment (rnuernberg) [reply src/CoreBundle/UniversalUploader/Controller/UploaderController.php:80]: because we have DI here - so we can access the original symfony service.

File: src/CoreBundle/private/core/TCMSModelBase.class.php [hunk 21]
```diff
@@ -102,7 +103,16 @@ protected function OutPutAjaxCallResult($functionResult)
...
+        /** @var SecurityHelperAccess $securityHelper */
+        $securityHelper = ServiceLocator::get(SecurityHelperAccess::class);
+
+        $user = $securityHelper->getUser();
+        if (null !== $user) {
+            $userObject = TdbCmsUser::GetNewInstance();
+            $userObject->Load($user->getId());
+        }
+
+        $this->data['oCMSUser'] = $userObject;
```
[the remark this answers is not in the document]
Comment (rnuernberg) [reply src/CoreBundle/private/core/TCMSModelBase.class.php:115]: fair

File: src/CoreBundle/private/core/TGlobal.class.php [hunk 22]
```diff
@@ -24,31 +25,6 @@ public static function setMode($mode)
         self::$mode = $mode;
     }
 
-    public function __get($sParameterName)
```
[the remark this answers is not in the document]
Comment (rnuernberg) [reply src/CoreBundle/private/core/TGlobal.class.php:27]: I am sure. I removed all usages I could find - and the magic should have been removed a long time ago

File: src/CoreBundle/private/library/classes/TCMSFields/TCMSField.class.php [hunk 23]
```diff
@@ -208,13 +210,9 @@ public function GetDisplayType()
...
-            $global = $this->getGlobal();
-            $fieldGroups = $this->oDefinition->GetPermissionGroups();
-            if ($global->oUser
-                && $global->oUser->oAccessManager
-                && $global->oUser->oAccessManager->user
-                && !$global->oUser->oAccessManager->user->IsInGroups($fieldGroups)
-            ) {
+            /** @var SecurityHelperAccess $securityHelper */
+            $securityHelper = ServiceLocator::get(SecurityHelperAccess::class);
+            if (false === $securityHelper->isGranted(CmsPermissionAttributeConstants::ACCESS, $this->oDefinition)) {
```
[the remark this answers is not in the document]
Comment (rnuernberg) [reply src/CoreBundle/private/library/classes/TCMSFields/TCMSField.class.php:215]: yes - it is a field definition - which implements `RestrictedByCmsGroupInterface` via `\ChameleonSystem\SecurityBundle\Bridge\Chameleon\TableObject\CmsFieldConf`

File: src/CoreBundle/private/library/classes/TCMSFields/TCMSFieldPropertyTable.class.php [hunk 24]
```diff
@@ -269,8 +272,9 @@ public function GetDisplayType()
         if ('1' == $this->oDefinition->sqlData['restrict_to_groups']) {
             // check if the user is in one of the connected groups
 
-            $fieldGroups = $this->oDefinition->GetPermissionGroups();
-            if (!$oGlobal->oUser->oAccessManager->user->IsInGroups($fieldGroups)) {
+            /** @var SecurityHelperAccess $securityHelper */
+            $securityHelper = ServiceLocator::get(SecurityHelperAccess::class);
+            if (!$securityHelper->isGranted(CmsPermissionAttributeConstants::ACCESS, $this->oDefinition)) {
```
[the remark this answers is not in the document]
Comment (rnuernberg) [reply src/CoreBundle/private/library/classes/TCMSFields/TCMSFieldPropertyTable.class.php:277]: yes

File: src/CoreBundle/private/library/classes/TCMSFields/TCMSFieldWYSIWYG.class.php [hunk 25]
```diff
@@ -208,8 +212,11 @@ private function convertToolbar($aToolbar)
     protected function getToolbar()
     {
         $aToolbar = $this->getCkeditorConfigProvider()->getToolbar();
-        $oUser = TCMSUser::GetActiveUser();
-        $aToolbar = $this->getModifiedToolbarByUser($oUser, $aToolbar);
+        /** @var SecurityHelperAccess $securityHelper */
+        $securityHelper = ServiceLocator::get(SecurityHelperAccess::class);
```
[the remark this answers is not in the document]
Comment (rnuernberg) [reply src/CoreBundle/private/library/classes/TCMSFields/TCMSFieldWYSIWYG.class.php:216]: removed

File: src/CoreBundle/private/library/classes/TCMSTableEditor/TCMSTableEditorCMSUser.class.php [hunk 26]
```diff
@@ -23,7 +26,7 @@ class TCMSTableEditorCMSUser extends TCMSTableEditor
     public function DefineInterface()
     {
         parent::DefineInterface();
-        $externalFunctions = array('CopyUserRights', 'ActivateUser', 'SwitchToUser');
```
[the remark this answers is not in the document]
Comment (rnuernberg) [reply src/CoreBundle/private/library/classes/TCMSTableEditor/TCMSTableEditorCMSUser.class.php:26]: was replaced by the impersonate user function from symfony (which is handelt via URL)

File: src/CoreBundle/private/library/classes/TCMSTableEditor/TCMSTableEditorCMSUser.class.php [hunk 27]
```diff
@@ -103,18 +103,23 @@ protected function GetCustomMenuItems()
...
     }
-
+    private function getUrlUtil(): UrlUtil
+    {
+        return ServiceLocator::get('chameleon_system_core.util.url');
+    }
     /**
      * @return bool
      */
     private function isSwitchToUserAllowed()
```
[the remark this answers is not in the document]
Comment (rnuernberg) [reply src/CoreBundle/private/library/classes/TCMSTableEditor/TCMSTableEditorCMSUser.class.php:113]: you are... :) see `\TCMSTableEditorCMSUser::GetCustomMenuItems`

File: src/CoreBundle/private/library/classes/TCMSTableEditor/TCMSTableEditorEndPoint.class.php [hunk 28]
```diff
@@ -674,9 +640,12 @@ public function GetPreviewURL()
...
         $portalId = $page->GetPortal()->id;
         $domainName = $this->getPortalDomainService()->getPrimaryDomain($portalId)->GetActiveDomainName();
-        $editLanguageID = TCMSUser::GetActiveUser()->GetCurrentEditLanguageID();
+        /** @var BackendSessionInterface $backendSession */
+        $backendSession = ServiceLocator::get('chameleon_system_cms_backend.backend_session');
+
+        $editLanguageID = $backendSession->getCurrentEditLanguageId();
 
-        return $this->getCurrentRequest()->getScheme().'://'.$domainName.'/'.PATH_CUSTOMER_FRAMEWORK_CONTROLLER.'?pagedef='.$this->sId.'&esdisablelinks=true&__previewmode=true&previewLanguageId='.$editLanguageID;
+        return $this->getCurrentRequest()?->getScheme().'://'.$domainName.'/'.PATH_CUSTOMER_FRAMEWORK_CONTROLLER.'?pagedef='.$this->sId.'&esdisablelinks=true&__previewmode=true&previewLanguageId='.$editLanguageID;
```
[the remark this answers is not in the document]
Comment (rnuernberg) [reply src/CoreBundle/private/library/classes/TCMSTableEditor/TCMSTableEditorEndPoint.class.php:648]: changed

File: src/CoreBundle/private/library/classes/TCMSTableEditor/TCMSTableEditorPortal.class.php [hunk 29]
```diff
@@ -185,32 +193,42 @@ protected function OnAfterCopy()
...
     }
 
     /**
-     * @param TCMSUser $oUser
      *
      * @throws ErrorException
      * @throws TPkgCmsException_Log
      */
-    protected function linkPortalToUser(TCMSUser $oUser)
+    protected function linkPortalToUser(?string $userId)
```
[the remark this answers is not in the document]
Comment (rnuernberg) [reply src/CoreBundle/private/library/classes/TCMSTableEditor/TCMSTableEditorPortal.class.php:215]: yes. documented in UPGRADE-7.2

File: src/CoreBundle/private/library/classes/dbobjects/TCMSContentBoxItem.class.php (hunk 1 above)
Comment (rnuernberg) [reply src/CoreBundle/private/library/classes/dbobjects/TCMSContentBoxItem.class.php:83]: yes - via `\ChameleonSystem\SecurityBundle\Bridge\Chameleon\TableObject\CmsModule`

File: src/CoreBundle/private/library/classes/dbobjects/TCMSUser.class.php (hunk 2 above)
Comment (rnuernberg) [reply src/CoreBundle/private/library/classes/dbobjects/TCMSUser.class.php:285]: jup. It will be 8.0

File: src/CoreBundle/private/modules/CMSInterface/CMSInterface.class.php (hunk 3 above)
Comment (rnuernberg) [reply src/CoreBundle/private/modules/CMSInterface/CMSInterface.class.php:38]: because I was thinking roles, not groups (every user has the ROLE_CMS_USER with no real id). removed

File: src/CoreBundle/private/modules/CMSModulePageTreeRPC/CMSModulePageTreeRPC.class.php (hunk 4 above)
Comment (rnuernberg) [reply src/CoreBundle/private/modules/CMSModulePageTreeRPC/CMSModulePageTreeRPC.class.php:121]: yes. see line 117. The entries are quoted. but there was a mistake. I needed to remove the `'` the follow and prefix the brackets

File: src/CoreBundle/private/modules/MTHeader/views/breadcrumb.view.php (hunk 5 above)
Comment (rnuernberg) [reply src/CoreBundle/private/modules/MTHeader/views/breadcrumb.view.php:2]: because we can always show the breadcrumb - it is not part of the login page anymore... (and we no longer have oUser)

File: src/CoreBundle/private/modules/MTTableEditor/MTTableEditor.class.php (hunk 6 above)
Comment (rnuernberg) [reply src/CoreBundle/private/modules/MTTableEditor/MTTableEditor.class.php:205]: /o\ changed

File: src/SecurityBundle/src/CmsUser/CmsUserModel.php (hunk 7 above)
Comment (rnuernberg) [reply src/SecurityBundle/src/CmsUser/CmsUserModel.php:26]: not sure I like to lose my getter...

File: src/CmsBackendBundle/src/ChameleonSystemCmsBackendBundle.php (hunk 9 above)
[the remark this answers is not in the document]
Comment (bestform) [reply src/CmsBackendBundle/src/ChameleonSystemCmsBackendBundle.php:9]: I do not see where this is used in the screenshot. Am I blind? :sweat_smile:

File: src/CoreBundle/DataModel/TableConfigurationDataModel.php [hunk 30]
```diff
@@ -0,0 +1,37 @@
...
+namespace ChameleonSystem\CoreBundle\DataModel;
+
+class TableConfigurationDataModel
+{
+    public function __construct(
+        readonly private string $id,
+        readonly private string $name,
+        readonly private string $cmsUsergroupId
+    ) {
+    }
```
[the remark this answers is not in the document]
Comment (bestform) [reply src/CoreBundle/DataModel/TableConfigurationDataModel.php:12]: I am pretty sure you want to change that. I won't fight over it but the getters do not make much sense. I know, countless years of having to use them for this purpose makes it feel wrong to avoid them now, but they really are not needed anymore. :)

File: src/CoreBundle/private/library/classes/TCMSFields/TCMSField.class.php (hunk 23 above)
[the remark this answers is not in the document]
Comment (bestform) [reply src/CoreBundle/private/library/classes/TCMSFields/TCMSField.class.php:215]: :+1:

File: src/SecurityBundle/src/CmsUser/CmsUserModel.php (hunk 7 above)
Comment (bestform) [reply src/SecurityBundle/src/CmsUser/CmsUserModel.php:26]: You do ;)

File: src/CmsBackendBundle/src/ChameleonSystemCmsBackendBundle.php (hunk 9 above)
[the remark this answers is not in the document]
Comment (rnuernberg) [reply src/CmsBackendBundle/src/ChameleonSystemCmsBackendBundle.php:9]: When creating the cmsUser (DataAccess in the same namespace as the user).

[the remark this answers is not in the document]
Comment (rnuernberg) [reply src/CmsBackendBundle/src/ChameleonSystemCmsBackendBundle.php:9]: sorry - mein komments were ment for https://github.com/chameleon-system/chameleon-base/pull/675#discussion_r1068225629

getPath is used by every bundle. I am use the more modern bundle structure here - the bundle class is in `src` - on level above the bundle path. If I omit the method, symfony is unable to locate a path of the form `@ChameleonSystemSecurityBundle/config/security.yaml`

File: src/CmsBackendBundle/src/BackendSession/BackendSession.php [hunk 31]
```diff
@@ -0,0 +1,124 @@
...
+
+        /** @var CmsUserModel $user */
+        $user = $this->security->getUser();
+        if (null === $user) {
+            return null;
+        }
+
+        $currentEditLanguage = $user->getCurrentEditLanguageIsoCode();
+
+        if (null !== $currentEditLanguage) {
```
[the remark this answers is not in the document]
Comment (rnuernberg) [reply src/CmsBackendBundle/src/BackendSession/BackendSession.php:60]: When creating the class in the data access, we ensure that currentEditLanguage is null if it was set to `''` in the database

File: src/CoreBundle/DataModel/TableConfigurationDataModel.php (hunk 30 above)
[the remark this answers is not in the document]
Comment (rnuernberg) [reply src/CoreBundle/DataModel/TableConfigurationDataModel.php:12]: I am not so sure. Currently, I am getting a nice code hint when using the getter. I have no idea how to ensure that I still get that when accessing the public property.

File: src/CoreBundle/private/library/classes/TCMSTableEditor/TCMSTableEditorPortal.class.php (hunk 29 above)
[the remark this answers is not in the document]
Comment (rnuernberg) [reply src/CoreBundle/private/library/classes/TCMSTableEditor/TCMSTableEditorPortal.class.php:215]: ok. moved

File: src/SecurityBundle/src/CmsUser/CmsUserModel.php (hunk 7 above)
Comment (rnuernberg) [reply src/SecurityBundle/src/CmsUser/CmsUserModel.php:26]: If you know how to add the proper code hint, then yes. Otherwise, I prefer getters for now

File: src/CoreBundle/DataModel/TableConfigurationDataModel.php (hunk 30 above)
[the remark this answers is not in the document]
Comment (bestform) [reply src/CoreBundle/DataModel/TableConfigurationDataModel.php:12]: It should work for everything except typed arrays - so basically every type annotation that is supported by php.

If you need the annotation for types in arrays you could annotate them like this:

```php
final class TestClass {

    public function __construct(
        /** @var array<string, string> */
        public readonly array $arr
    ){}

}
```

Would this work for you?

File: src/SecurityBundle/src/CmsUser/CmsUserModel.php (hunk 7 above)
Comment (bestform) [reply src/SecurityBundle/src/CmsUser/CmsUserModel.php:26]: see above

File: src/CoreBundle/DataModel/TableConfigurationDataModel.php (hunk 30 above)
[the remark this answers is not in the document]
Comment (rnuernberg) [reply src/CoreBundle/DataModel/TableConfigurationDataModel.php:12]: yes - and no. I did change `\ChameleonSystem\CoreBundle\DataModel\TableConfigurationDataModel` to use public readonly. But the CmsUserModel implements `UserInterface, PasswordAuthenticatedUserInterface` - which dictate that it has methods such as getRoles, getPassword, getSalt and getUsername. I find it very strange to use getters for some properties but not for others - so in this case I am keeping it

File: src/SecurityBundle/src/CmsUser/CmsUserModel.php (hunk 7 above)
Comment (rnuernberg) [reply src/SecurityBundle/src/CmsUser/CmsUserModel.php:26]: see above :)

File: src/CoreBundle/DataModel/TableConfigurationDataModel.php (hunk 30 above)
[the remark this answers is not in the document]
Comment (bestform) [reply src/CoreBundle/DataModel/TableConfigurationDataModel.php:12]: that's a good point.
